vehicle/src/Vehicle/Backend/Isabelle/Compile.hs

```diff
@@ -0,0 +1,897 @@
+    if isAnnotatedAsProperty anns
+      then return "" -- Done via localeAssm gathering
+      else compileFunDef localeAssms n t binders body
+        --bindersT bindersV cbody
+  DefRecord _ n _ _t fs -> do
+    -- t' <- compileExpr False localeAssms t
+    fs' <- traverseRecordFields (compileExpr False localeAssms) fs
+    return $
+      "record"
+        <+> compileIdentifier n
+        <+> "="
+        <> line
+        <> indent 2 ((vsep :: [Code] -> Code) $ fmap (\(field, fieldType) -> pretty field <+> "::" <+> "\""<>fieldType<>"\"") fs')
+
+-- | Compile a 'network' declaration
+compilePostulate :: Code -> Code -> LocaleDef
+compilePostulate name t = (NetworkDefStatement name t)
+
+compileExprUnfoldings :: Expr DecidabilityBuiltin -> [Code]
+compileExprUnfoldings expr = case expr of
+  App fun args -> concatMap compileExprUnfoldings (fun : (fmap argExpr (NonEmpty.toList args)))
+  Lam _ binder body -> compileExprUnfoldings (typeOf binder) ++ compileExprUnfoldings body
+  Pi _ binder result -> compileExprUnfoldings (typeOf binder) ++ compileExprUnfoldings result
+  Let _ bound binder body -> compileExprUnfoldings bound ++ compileExprUnfoldings (typeOf binder) ++ compileExprUnfoldings body
+  Record _ _ fs -> concatMap (compileExprUnfoldings . snd) fs
+  RecordAcc _ r _ -> compileExprUnfoldings r
+  Builtin _ _ -> []
+  FreeVar _ n -> ["unfolding " <> pretty (nameOf n) <>"_def"]
+  BoundVar _ _ -> []
+  Universe _ _ -> []
+  Meta _ _ -> []
+  Hole _ _ -> []
+
+compileExpr :: (MonadIsabelleCompile m) => Bool -> [LocaleDef] -> Expr DecidabilityBuiltin -> m Code
+compileExpr isOutType localeAssms expr = do
+  logEntry expr
+  result <- case expr of
+    Hole {} -> resolutionError currentPhase "Hole"
+    Meta {} -> resolutionError currentPhase "Meta"
+    Universe _ l -> developerError $ "compilation of universes to Isabelle unsupported: " <> pretty l
+    FreeVar _ n -> return $ annotateConstant [] (pretty (nameOf n))
+    BoundVar p ix -> do
+      n <- ixToProperName p ix
+      return $ annotateConstant [] (pretty n)
+    Pi _ binder result -> case binderNamingForm binder of
+      OnlyType -> do
+        cInput <- compileBinder localeAssms binder
+        cOutput <- addNameToContext binder $ compileExpr True localeAssms result
+        return $ annotate ([], 99) $ cInput <+> "\\<Rightarrow>" <+> cOutput
+      _ -> do
+        let (binders, body) = foldBinders PiBinder binder result
+        compileTypeLevelQuantifier localeAssms Forall (binder :| binders) body
+    Let _ bound binder body -> do
+      cBoundExpr <- compileLetBinder localeAssms (binder, bound)
+      cBody <- addNameToContext binder $ compileExpr False localeAssms body
+      return $ "let" <+> cBoundExpr <+> "in" <+> cBody
+    Lam _ binder body -> compileLam localeAssms binder body
+    Builtin _p b -> compileBuiltin isOutType localeAssms b []
+    App fun args -> compileApp isOutType localeAssms fun args
+    Record _p _i fs -> do
```

Comment:
This is quite complicated. Can we split it out into a `compileRecord` function?

Reply:
done